src/messengers/vk/vk/updates.go

```diff
@@ -0,0 +1,226 @@
+	"golang.org/x/time/rate"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"log"
+	"path/filepath"
+	"strings"
+
+	msg "github.com/Pelmenner/TransferBot/proto/messenger"
+	"github.com/SevereCloud/vksdk/v2/api"
+	"github.com/SevereCloud/vksdk/v2/object"
+)
+
+func (m *Messenger) Run(ctx context.Context) {
+	restartLimiter := rate.NewLimiter(rate.Limit(Config.LongPollRestartMaxRate), 1)
+	for {
+		if err := restartLimiter.Wait(ctx); err != nil {
+			log.Printf("error waiting for limiter: %v", err)
+		}
+		if err := m.longPoll.Run(); err != nil {
+			log.Print("VK longpoll error:", err)
+		}
+	}
+}
+
+func (m *Messenger) processMessage(message object.MessagesMessage, chat *msg.Chat) error {
+	if message.IsCropped {
+		message = m.getFullMessage(message)
+	}
+	if err := m.processCommand(message, chat); err != nil && err != errCommandNotFound {
+		return err
+	}
+	standardMessage := msg.Message{
+		Text: message.Text,
+		Sender: &msg.Sender{
+			Name: m.getSenderName(message),
+			Chat: &msg.Chat{Name: "vk"},
+		},
+	}
+	var walls []*object.WallWallpost
+	for _, attachment := range message.Attachments {
+		switch attachment.Type {
+		case "photo":
+			standardMessage.Attachments = m.processPhoto(attachment.Photo, chat.Id, standardMessage.Attachments)
+		case "wall":
+			walls = append(walls, &attachment.Wall)
+		case "doc":
+			standardMessage.Attachments = m.processDocument(attachment.Doc, chat.Id, standardMessage.Attachments)
+		}
+	}
+	if message.ReplyMessage != nil {
+		standardMessage.Text += "\nin reply to..."
+	}
+	err := m.MessageCallback(&standardMessage, chat)
+	for _, wall := range walls {
+		if err = m.processWall(*wall, chat); err != nil {
+			return err
+		}
+	}
+	if message.ReplyMessage != nil {
+		if err = m.processMessage(*message.ReplyMessage, chat); err != nil {
+			return err
+		}
+	}
+	for _, message := range message.FwdMessages {
+		_ = m.processMessage(message, chat)
+	}
+	return nil
+}
+
+// Returns not cropped message requesting it by id (extracted from given message)
+func (m *Messenger) getFullMessage(message object.MessagesMessage) object.MessagesMessage {
+	messageResponse, err := m.vk.MessagesGetByConversationMessageID(api.Params{
+		"conversation_message_ids": message.ConversationMessageID,
+		"peer_id":                  message.PeerID,
+	})
+	if err != nil || messageResponse.Count == 0 {
+		log.Print("could not get vk message by conversation id ", err)
+		return message
+	}
+	return messageResponse.Items[0]
+}
+
+func (m *Messenger) getSenderName(message object.MessagesMessage) string {
+	userResponse, err := m.vk.UsersGet(api.Params{"user_ids": message.FromID})
+	if err != nil || len(userResponse) == 0 {
+		log.Print("could not find vk user with id ", message.FromID, err)
+		return ""
+	}
+	// have not found a way to get chat name
+	return userResponse[0].FirstName + " " + userResponse[0].LastName
+}
+
+var errCommandNotFound = fmt.Errorf("command not found")
```

Comment:
a bit strange to define an error literally in the middle of nowhere

Reply:
It is right about the function that returns it (`processCommand`). It has to be outside to be available in other functions (particularly in `processMessage`, which calls `processCommand`)